locator/abstract_replication_strategy.cc

```diff
@@ -315,13 +315,74 @@ static const auto default_replication_map_key = dht::token::from_int64(0);
 
 future<mutable_vnode_effective_replication_map_ptr> calculate_effective_replication_map(replication_strategy_ptr rs, token_metadata_ptr tmptr) {
     replication_map replication_map;
+    ring_mapping pending_endpoints;
+    ring_mapping read_endpoints;
     const auto depend_on_token = rs->natural_endpoints_depend_on_token();
     const auto& sorted_tokens = tmptr->sorted_tokens();
     replication_map.reserve(depend_on_token ? sorted_tokens.size() : 1);
-    if (depend_on_token) {
+    if (const auto& topology_changes = tmptr->get_topology_change_info(); topology_changes) {
+        const auto& all_tokens = topology_changes->all_tokens;
+        const auto& base_token_metadata = topology_changes->base_token_metadata
+            ? *topology_changes->base_token_metadata
+            : *tmptr;
+        const auto& current_tokens = tmptr->get_token_to_endpoint();
+        for (size_t i = 0, size = all_tokens.size(); i < size; ++i) {
+            co_await coroutine::maybe_yield();
+
+            const auto token = all_tokens[i];
+
+            auto current_endpoints = co_await rs->calculate_natural_endpoints(token, base_token_metadata);
+            auto target_endpoints = co_await rs->calculate_natural_endpoints(token, topology_changes->target_token_metadata);
+
+            auto add_mapping = [&](ring_mapping& target, std::unordered_set<inet_address>&& endpoints) {
+                using interval = ring_mapping::interval_type;
+                if (!depend_on_token) {
+                    target += std::make_pair(
+                        interval::open(dht::minimum_token(), dht::maximum_token()),
```

Comment:
Why does this work?

Reply:
> Why does this work?

`pending_endpoints/read_endpoints` maps ranges to corresponding endpoints, we need this map since different ranges could map to different endpoints based on the relation between natural endpoints for a given token. In case `!depend_on_token` all ranges will behave in the same way, so we can map them all by single `(-inf, +inf)` key.